### ISSUE TITLE
Querion how to chain with different left type

### ISSUE THREAD

**Issue by edgarjrg** [fluture-js/Fluture#371]
Hi ! 

This tests is successful for me
```
test('chain with different left type', () => {
    fork(x => expect(x).toEqual(3))(fail)(chain<boolean, any, any>(x => reject(false))(reject(3)))
  })
```
However i have to ignore the types otherwise i would get
```
Argument of type 'FutureInstance<number, any>' is not assignable to parameter of type 'FutureInstance<boolean, any>'.
  Type 'number' is not assignable to type 'boolean'.ts(2345)
```
What is the right method that has this type signature
```ts
export function chain?<LA, RA, LB, RB>(mapper: (value: RA) => FutureInstance<LB, RB>): (source: FutureInstance<LA, RA>) => FutureInstance<LB, RB>

export function chainRej?<LA, RA, LB, RB>(mapper: (value: LA) => FutureInstance<LB, RB>): (source: FutureInstance<LA, RA>) => FutureInstance<LB, RB>
```

That's pretty much how i chain `Promises`, but maybe Futures are supposed to be though or combined differently.

**Comment by edgarjrg** [comment]
As it may have not applied the `mapper` at all if rejected, actually i think it may be more like this:

```ts
export function chain?<LA, RA, LB, RB>(mapper: (value: RA) => FutureInstance<LB, RB>): (source: FutureInstance<LA, RA>) => FutureInstance<LA, never> | FutureInstance<LB, RB>
```
That resolves to 
```ts
export function chain?<LA, RA, LB, RB>(mapper: (value: RA) => FutureInstance<LB, RB>): (source: FutureInstance<LA, RA>) => FutureInstance<LA | LB, RB>
```

Promises never deal with how the error is typed as is always `any`, but actually i think this is the more accurate representation, but again, my lack of knowledge of the Future type might be in the way. I've yet not seen any library that deals with different types in the left branch, is always one generic type `E` that never changes.

**Comment by Avaq** [comment]
> how to chain with different left type

It is not allowed to `chain` two different types of Future: `(Future Number) a` is not the same Monad as `(Future Boolean) a`.

The problem is that without Type Checking, we can never ensure this constraint using just plain JavaScript. So only TypeScript users are aware of the constraint.

The solution is to make the types match before you `chain`:

```js
const boolToInt = b => b ? 0 : 1;

fork (x => expect(x).toEqual(3))
     (fail)
     (chain(x => mapRej(boolToInt)(reject(false)))
            (reject(3)))
```

...or, if you cannot make the types match, then don't `chain`, but `map`:

```js
fork (x => expect(x).toEqual(3))
     (fork (fail) (fail))
     (map(x => reject(false))
            (reject(3)))
```

**Comment by Avaq** [comment]
Note the type signature also reflects this rule:

```hs
chain :: Monad m => (a -> m b) -> m a -> m b
```

Which, specialized in Future becomes the following. Note that I have replaced `m` with `Future e`:

```hs
chain :: (a -> (Future e) b) -> (Future e) a -> (Future e) b
```

This shows that the type of `e` cannot change between the two Futures.

**Comment by edgarjrg** [comment]
Thank you for answering so fast! I see now 🎉 

TL;DR;

What i'm trying to do i think, can only be achieved with `(Future string | number) a`



I need to express the behavior of calling sequentially different Futures that each return their own type of errors, so calling each would hide the previous left and right. 

Its something very similar to `and` and i would use that if it would give me the result of the previous Future, but it does not, so i'm not sure how to get it.

I simplified my use case and made this example:
```ts
type L1 = boolean
type R1 = string

type L2 = string[]
type R2 = string

type L3 = number
type R3 = number

type L4 = {
    code: 'Fail 1'
} | {
    code: 'Fail 2'
} | {
    code: 'Fail 3'
}

type R4 = 'finish'

function getFromSource1(_s: string): FutureInstance<L1, R1> {
    return reject(false)
}

function getFromSource2(_s: string): FutureInstance<L2, R2> {
    return resolve('made it')
}

function getFromSource3(_s: string): FutureInstance<L3, R3> {
    return reject(404)
}

function getFromMultipleSources(s: string): FutureInstance<L4, R4> {
    return S.pipe([
        S.pipe([
            getFromSource1,
            mapRej(() => ({ code: 'Fail 1' as 'Fail 1' })),
        ]),
        chain(
            S.pipe([
                getFromSource2,
                mapRej(() => ({ code: 'Fail 2' as 'Fail 2' }))
            ])
        ),
        chain(
            S.pipe([
                getFromSource3,
                bimap(
                    () => ({ code: 'Fail 3' as 'Fail 3' }),
                    () => 'finish' as 'finish'
                )
            ])
        ),
    ])(s)

}
```
i'm assigning a different type to each failed step, as much as i could simply not and made 
```ts
type L5 = L1 | L2 | L3

function getFromMultipleSources2(s: string): FutureInstance<L5, R4> {
    return S.pipe([
        getFromSource1,
        chain(getFromSource2),
        chain(getFromSource3),
        map(() => 'finish' as 'finish')
    ])(s)

}

```
this does not changes the nature that each `getFromSource3` returns their own `L` and `R` pair, and i want to reflect that in `getFromMultipleSources(2)`, to callers to be able to map the failures. In Promise world i would have done

```ts
function getFromSource4(_s: string): Promise<R1> {
    return Promise.reject(false)
}

function getFromSource5(_s: string): Promise<R2> {
    return Promise.resolve('made it')
}

function getFromSource6(_s: string): Promise<R3> {
    return Promise.reject(404)
}

function getFromMultipleSourcesP(s: string): Promise<R4> {
    return getFromSource4(s).catch(() => Promise.reject({ code: 'Fail 1' }))
        .then(r => getFromSource5(r).catch(() => Promise.reject({ code: 'Fail 2' })))
        .then(r => getFromSource6(r).catch(() => Promise.reject({ code: 'Fail 3' })))
        .then(() => 'finish')
}
```
but the Promise`reason` is always `any` :(

At first, when i identify any failure i would send it to the right, thats better suited to handle different types, then i would finish with something like `FutureInstance<never, L4 | R4>` some kind of "unfailable Future" thats some concept worth exploring, but it did piping more difficult because L4 and R4 are very different in shape.

What i'm doing is extrange? i've never had any problems in Promise world because `L` is `any`. In Fluture world i've been trying to get the juice out of typescript, but the types make it hacky, even though its looks like it works just fine to do `chain(x => reject(false))(reject(3))`.

I've been having problems with the `L` because i want each `chain?` to union itself, else it would say that the current type of `L` is not assignable previous type `L` ('Fail 1' !== 'Fail 2'), but if i do

```ts
function response(code: 'Fail 1' | 'Fail 2' | 'Fail 3'): L4 {
    return {
        code
    }
}

function getFromMultipleSources3(s: string): FutureInstance<L4, R4> {
    return S.pipe([
        S.pipe([
            getFromSource1,
            mapRej(() => response('Fail 1')),
        ]),
        chain(
            S.pipe([
                getFromSource2,
                mapRej(() => response('Fail 2'))
            ])
        ),
        chain(
            S.pipe([
                getFromSource3,
                bimap(
                    () => response('Fail 3'),
                    () => 'finish' as 'finish'
                )
            ])
        ),
    ])(s)

}
```

Typescript is not capable of identify that `{ code: 'Fail 1'}` and `{ code: 'Fail 2'}` are members of L4. however with `response` everything in `L` is of `L4`.



I wanted to made it more readable with `S.pipe` but it makes typescript not complain at all, in reality i'm using `fp-ts/lib/pipeable/pipe`. With fp-ts pipe or Fluture.pipe or Fluture.chain the typescript compilers warns everything.

Thank you for your help, i'm just posting this long to help anybody that might need the reasoning.
